### ISSUE TITLE
Proposal: Entitlements in Moby

### ISSUE THREAD

**Issue by n4ss** [moby/moby#32801]
# Entitlements in Moby

This issue captures a draft design proposal for an entitlement mechanism that can be leveraged by moby to describe what additional permissions a specific service should be allowed to have when executing in a swarm.

An entitlement is a single right granted to a particular service/container that gives it additional permissions above and beyond what it would ordinarily have. An entitlement is a piece of configuration information included in the service spec, telling the container engine that executes a service allow access to certain resources or perform certain operations. In effect, an entitlement extends the sandbox and capabilities of your service to allow a particular operation to occur.

The Docker cli currently supports over 100 command-line flags. By implementing an entitlement mechanism we plan to allow downstream consumes of moby such as docker to unify all of the security related flags into a single mechanism that is granular enough to be useful, platform independent, and understandable by a non-expert user. This mechanism can be seen as Moby's equivalent of Apple’s App Store permission model, where apps are granted capabilities to operate beyond the normal privileged of an application, such as access to the keychain or the ability of enabling push-notifications.

## Goal

The goal is to simplify the way downstream users ask for permissions for their containers/services.

In order for a service to use a specific entitlement, access to that entitlement has to be granted. The objective is to have a grant mechanism that would look like this:

Specifying a full list of grants:
> docker service create --grants=[entitlement] alpine

Adding a grant to a service
> docker service create --add-grant=[entitlement] alpine

Removing a grant from a service
> docker service create --rm-grant=[entitlement] alpine
> 
## Current proposal for entitlements

|Entitlement|Privileges|Capabilities|Blocked syscalls|On Windows|
|:-:|:-:|:-:|:-:|:-:|
|engine.api| Full access to the Docker API.||||
|host.devices.none| Read only on /sys/ mount; Read only flag for cgroup mount; No host device in /dev; No read/write on /proc/kcore||| N/A |
|host.processes.view|Effectively --pid=host||| N/A |
|host.devices.view| Activate userns? ||| N/A |
|host.devices.mount|Add SYS_ADMIN and allow a device to be mounted in.||| N/A |
|network.none| No access to /proc/pid/net, /proc/sys/net; No access to /sys/class/net| No NET_ADMIN, NET_BIND_SERVICE, NET_RAW, NET_BROADCAST| socket, socketpair, setsockopt, getsockopt, getsockname, getpeername, bind, listen, accept, accept4, connect, shutdown,recvfrom, recvmsg, sendto, sendmsg, sendmmsg, sethostname, setdomainname, bpf||
|network.user| |CAP_NET_RAW, CAP_NET_BIND_SERVICE?, CAP_NET_BROADCAST?|sethostname, setdomainname bpf, setsockopt(SO_DEBUG)||
|network.proxy||Add: CAP_NET_RAW, CAP_NET_BROADCAST, CAP_NET_BIND_SERVICE, CAP_NET_BIND_SERVICE, CAP_NET_RAW|||
|network.admin||CAP_NET_ADMIN, CAP_NET_BROADCAST, CAP_NET_BIND_SERVICE, CAP_NET_RAW|||
|security.none|Block access to sentitive paths: /sys/kernel/security , /sys/kernel/debug (ftrace), /sys/kernel/livepatch, /sys/fs/selinux, /sys/fs/cgroup, debugfs, securityfs, selinuxfs, /proc/sys/kernel/, /proc/config.gz, /boot, /proc/{mem,cpu,kcore,kmem,sysrq-trigger,bus} <br> No MAC/DAC policy read/write or configuration/state change <br> NoNewPrivileges activated|Drop:No CAP_MAC_\*, CAP_DAC_\*, CAP_SETPCAP, SYS_PTRACE, CAP_SET_\*, CAP_FSETID, CAP_SYS_ADMIN|bpf, ptrace, seccomp, arch_prctl, personality, setuid/setgid?, madvise, prctl(PR_CAPBSET_DROP, PR_SET_\*, ..)||
|security.view|Read Only rights on sensitive filesystems / fs directories and MAC/DAC policies|Add:CAP_MAC_\*, CAP_DAC_\*, CAP_SETPCAP <br>Drop: CAP_LINUX_IMMUTABLE?, CAP_SET_\*, CAP_FSETID, , SYS_PTRACE, CAP_SYS_ADMIN|||
|security.admin||Add: CAP_MAC_\*, CAP_DAC_\*, CAP_LINUX_IMMUTABLE,  CAP_SYS_MODULE, CAP_SYS_PTRACE, CAP_SYSLOG,  CAP_FSETID, CAP_SYS_BOOT|||
|security.unconfined| Disable seccomp and LSMs ||||
|security.read-only| Mounts the container's filesystem as read-only||||
|resources.limit=`value`| `value` is a percentage of available resources in the container at launch context for: Pids, perf_event, blkio, hugetlb, freezer, net_cls, net_prio, cpuset, memory, systemd. Set ulimits properly||||
|resources.all|resources.limit=100|CAP_SYS_RESOURCES|||
|debug|security.unconfined|Add:CAP_SYS_ADMIN, CAP_SYS_PTRACE, CAP_SYSLOG|||

## Examples of usecases

Following examples are meant to show how to use entitlements for the most downloaded images on Docker Hub. They will probably be edited as new entitlements will probably be added and/or adjusted.

People used to run the following command in various use-cases:
> docker run --privileged imagename:label

We impact privileges by reducing the amount of rights granted, let’s see some examples:

People who need to use raw sockets to control link discovery and aggregation
Before:
 > docker run --privileged image:label

After:
 > docker run --grant=network.proxy image:label

Docker in Docker would probably look like:
Before:
 > docker run --privileged docker:dind
After:
 > docker run --grant=network.admin,host.devices.mount,engine.api,security.view docker:dind

Nginx:
> docker run --grant=network.admin nginx:latest

Redis:
> docker run --grant=resources.all,network.admin redis:latest

Httpd:
> docker run --grant=network.admin httpd:latest

Ubuntu / Alpine:
For network usage:
> docker run --grant=network.admin,security.view ubuntu:latest

Docker Registry:
> docker run --grant=network.admin registry:latest

MySQL:
> docker run --grant=network.admin,resources.all mysql:latest

We need examples for MySQL, Mongo, Logstash, ElasticSearch, Postgres, Node, Wordpress, Java, Jenkins and other widely used images.

Long term, you should be able to tie a set of entitlements to an image if you want to as a publisher so the more example, the better.

## Open Questions

Can the same privilege name mean different things for docker version to docker version? We would be treating entitlements as we treat default profiles.

Do all these entitlements also make sense on windows?

Should we provide a way to create custom entitlements?

## What we need from the community

- Validation that these entitlements are: high-level enough, non-overlapping and correctly implemented using the lower level primitives available from Linux
- Examples of use-cases where the entitlements fit/don't fit, and if they add too much privilege over what users are usually doing

**Comment by thaJeztah** [comment]
Wondering if we can add a column how these translate to native Windows containers; ideally we'd have entitlements that are generic enough to be applied to both Windows and Linux containers

**Comment by n4ss** [comment]
@thaJeztah done, I totally agree on that

**Comment by tych0** [comment]
I think we can drop `cpuacct` from the cgroups set, there's no values that really make sense to be writable there. I'll have a think about what a percentage value means for the others.

**Comment by tych0** [comment]
Another thought is about CAP_SYS_ADMIN. That contains a lot of stuff that is sort of unrelated to "sys admin-ing", like mounts and whatnot. This also restricts e.g. the ability to mount. debug may be a reasonable fit for this, but it's worth noting.

**Comment by lanmingle** [comment]
这个很好，觉得很适合 Moby 这个项目，希望能拆分模块，同时提供
 （Web）API 接口，设计成丰富的权限以便提供大部分服务长期使用。

**Comment by n4ss** [comment]
@tych0 
- `cpuacct`: dropped, agree that it's useless in that group
- `CAP_SYS_ADMIN`: sys-admins probably need `swap(on|off)`, `setdomainname` and other stuff in it too no? (I think this cap should be on for `security.admin` and `debug`)

**Comment by tych0** [comment]
So I was thinking about some of this today. r.e. devices:

Some of these privilege levels might correspond to access to particular devices (e.g. for network.none, there is no need to access `/dev/net/tun`, etc.). I wonder if we can't do something clever with the devices
cgroup here: by default, deny all access (and mknod) for devices except for common ones: /dev/null, /dev/zero, /dev/full, dev/random, etc.

Then we can make a list of devices (/dev/net/tun for network.proxy, etc.) to allow with a given entitlemen
t. More dangerous devices (/dev/mem, /dev/kmem) or devices that don't make sense (perhaps /dev/kvm, etc.)
are then disallowed by default via the devices cgroup.

The drawback is that sometimes users do weird things (I spoke with a guy at DockerCon recently who was using /dev/shm to do something fancy and was interested in using containers), so we perhaps need a way to add
 arbitrary major:minor devices to the allowed list if we do this, but I think it's probably worth doing.

I'm also not sure why `host.devices.view <=> --pid=host`. I assume host.devices.view is supposed to allow reading of stuff like /dev/sda, which seems unrelated to whether or not the container is in the host pid namespace.

**Comment by PatrickLang** [comment]
__Questions__

- What would an example usage for `resources.limit` ? 


__Areas where we definitely need answers for Windows containers__

1. Windows does have limits available for memory (bytes), cpu (core count & % per core), and disk (iops, size) but I don't see those listed in the capabilities above

2. I think the tech is there in Windows to build the `net_*` capabilities although I'm not sure if they're hooked up in moby today. @Kallie-b can you follow up with the network devs to get feedback on this design?

__Linux containers under Hyper-V Isolation__
`network.*` and `security.*` should be able to work for Linux containers running with the Hyper-V isolation that we recently announced since it's running a real Linux kernel in a minimal VM. That's still a WIP but I'll get the devs looking at it. Resources will be a bit different because some resource controls are done through the hypervisor in addition to the cgroup. `host.*` probably doesn't make sense since we won't be translating things like /dev/ttyS, GPU, ... due to differences in the driver models.

**Comment by thaJeztah** [comment]
@PatrickLang Good point; I'm not sure resource-limits is something that fits in the entitlements; setting that sound like in direct "conflict" with the existing `--memory-limit` / `--memory-reservation`, and not sure how that would affect scheduling

I _think_ entitlements are more what _capabilities_ a container or service has, but I can see the line between that and constraining resources is a bit blurry. We should define that line better

**Comment by n4ss** [comment]
@PatrickLang Sorry yes, I just added the Windows column, I'll add Windows limits and split Linux and Windows. Let's say you want to split the resources used by containers on the same host. You could be able to launch the first one (webserver for instance) with a `--resources.limit=40` for 40% of available resources and `--resources.limit=30` for 30% of available resources on the second one (database for example).

@tych0 done for the devices cgroups, I agree on the devices to allow/disallow based on level of privileges. I'll propose stuff regarding current categories, feel free to propose relevant use-cases and associated devices to provide access to.

**Comment by thaJeztah** [comment]
So what would the approach be for a container that needs processing power, and another container that needs memory? Putting all resource-limits under a single option does feel like a really "course" way to set those limits; I can imagine more fine-grained control being needed (limit cpu, network, disk-usage) instead of setting them all at once

**Comment by aschmitz** [comment]
A few questions:

1. Why does, say, MySQL require `network.admin`? Are there less-scary sets of permissions that would allow listening on a socket/port?
2. Can you make `*.none` the default (or perhaps "the default upon opting in to capabilities", whatever that means) rather than making users "add" it as a "capability" (when it's more of a restriction)? Similarly for things like `security.read-only`, where it's not really a capability. I think of capabilities as something you can "drop" later and be even *more* secure, although maybe you don't plan for that ability.
3. Can you rename `engine.api` to something scarier, like `docker/moby.full-control`?

Also, `resources.limit` feels kind of wedged in here. It would be useful, but I agree with @thaJeztah that there are likely to be cases where you want to have more control over individual resources. It doesn't feel like a "capability" thing, but if you're implementing it I would like to see some sort of absolute limits be able to be set as well ("2 GB RAM, 1 GHz CPU, 10 GB storage", for example, although percentages may be nice for scalable containers as well). I don't know how feasible that would be, but I'd break it out into a separate issue, as I'm sure it's tricky.

**Comment by diogomonica** [comment]
@aschmitz I agree that `*.none` would be great, but in practice what ends up happening is that right now a `docker run` has a set of default capabilities that you would have to always be verbose about adding.

Another option is to add a `default` that you always have to provide:

`docker service create --grant=default,network.admin redis`

Turning this into a least-privilege grant:

`docker service create --grant=network.admin redis`

Thoughts?

**Comment by aschmitz** [comment]
My concern there is that `default` then gets pretty locked in: if you later
want to remove a permission from it, you'll disrupt a bunch of users.

What about documenting the default set of capabilities, and defaulting to
none if any are specified, making users specify that they want each of the
default capabilities? This won't break capabilities users if you remove
defaults in the future, although it has the drawback of possibly making
some users hardcode the defaults without thinking about them.

**Comment by n4ss** [comment]
@thaJeztah We can of course improve the granularity for resources. I like the resource entitlement since it's part of sandboxing to control the amount of resources given to a sandboxed process/container/context. If we can leverage cgroups and CAP_SYS_RESOURCE in the set of grants that'd be great imho.

What's your opinion on main "categories" needed the most: `cpu`, `memory`, `network`? 

\
@aschmitz We indeed have _non-`none`_ permissions by default, but the idea behind those `none` is to provide a way to lock-down you container even more than usual.

We can document the default set of grants that we give (same as `seccomp` or `apparmor` default profiles) and let you improve the chose if you need more or less security for each category. The default entitlements we'll provide to the container should match the today's behaviors.

If we move out of the default set a permission to a different level for an entitlement, this should be treated as a security bugfix, same way as it would for a `seccomp` or `apparmor` default profile modification.

**Comment by kallie-b** [comment]
@PatrickLang -- yep, I'll follow up with my team on this :)

**Comment by RRAlex** [comment]
While I salute the idea of simplifying the model for new comers, I really hope that it will be possible to unbundle these entitlements or, at least, be able to create our own custom ones and specify a single capability to be added or removed if need be.

Bundling capabilities together is interesting for a quick and dirty security layer but is a very arbitrary & risky proposition if we don't get the granular control back for those who work in high risk environments, know what they need and and have the time to lock things down properly.

I'll probably repeat myself from other related issues, and sorry for being blunt, but trying to be platform agnostic at the cost of platform specific security is not a good direction.

**Comment by n4ss** [comment]
@RRAlex Agreed, we should provide a way to mark entitlements system as `disabled` for those who really need the granularity, I'll add that.

Creating custom entitlements should also be possible too indeed, in term of layout we could have the following:

**libEntitlements**
- provides an interface to register entitlements and their callbacks (format of entitlements, callbacks and constraints TBD)
- provides hooks to make a configuration action / permission decision based on registered entitlements and associated callbacks

If/How libEntitlements should setup/run the callback itself in the most secure manner, TBD too.

**Moby**
- has a set of default callbacks to register (based on easily manageable profile maybe?)
- has libEntitlements hooks calls for specific entitlements where needed to setup and manage resources, security and system configurations as well as APIs access.
- provides a way to opt out of entitlements (which would simply disable the hooks)
- allows user to provide custom entitlements and callbacks

**Comment by alfred-landrum** [comment]
If I understand the proposal, entitlements represent more than just a set of capabilities. Regarding the image creation side of this:
> Long term, you should be able to tie a set of entitlements to an image if you want to as a publisher so the more example, the better.

In the current proposal then, there would not be a way for an image creator to encode the entitlements a container needs to successfully run? How would that occur in the future?

**Comment by n4ss** [comment]
@alfred-landrum that's correct, we should leverage every tool available for every architecture in an user-friendly way.

I think that this proposal should focus on the definition of entitlements. But we should definitely add support in `docker-compose` for entitlements at first and longer term we should have a proper way to tie a security profile to an image (and have it part of the signed data) that publishers push to registries. Let's discuss it on a new proposal PR?

**Comment by briceburg** [comment]
Is an entitlement planned that will support containers requiring the `IPC_LOCK` capability to lock memory?

It's recommended for hashicorp vault (https://github.com/hashicorp/docker-vault/blob/9bbc74d04e5986f582fe08c2699c5e2ebfc9adc2/0.7.0/docker-entrypoint.sh#L85) and elasticsearch -- and I have a real-world use-case of running these as global services in independent (aka dedicated) swarms.

**Comment by thaJeztah** [comment]
For resources; possibly "reservations" make sense (e.g. this image requires 300MB of ram), but not constraints.

I propose to split the resource limits to a separate proposal

**Comment by n4ss** [comment]
@briceburg I'm thinking about where it might fit currently, or create entitlements under `resources.memory`

**Comment by mato** [comment]
@diogomonica:
> I agree that `*.none` would be great, but in practice what ends up happening is that right now a `docker run` has a set of default capabilities that you would have to always be verbose about adding.

Here's an idea: Could we grandfather in the current set of default capabilities for images _without an encoded profile_, and make `*.none` the default for images with an encoded profile?

This would give us a path towards "everything runs as least privilege by default" in the future which is a good thing. It does tie the use of entitlements to support for encoding them in images but I think it's worth it.

Further, it would also eliminate the need for eg. `--grant=security.none` which reads wrong to me -- it's not "granting" anything.

**Comment by n4ss** [comment]
@mato we have absolutely no way not to break compatibility that way. Maybe `--grant` is not the best name for this maybe `--rights` works best and would make more sense on `--rights=security.none`.

In my opinion, there is no way to have absolute least privilege here without breaking behavior of the CLI and having it extremely verbose for people afterward.

**Comment by mato** [comment]
@n4ss What I'm suggesting is that we keep compatibility with existing images (those without an encoded rights profile, i.e. everything on the hub today) and only enable the new least-privilege behaviour for images which have an encoded rights profile. So, by publishing an image with an encoded profile you (the publisher) opt in to the "least-privilege + rights" model.

From the user's / CLI point of view nothing would change for images without a rights profile. `--privileged` and all current options would continue to work.

For images with a rights profile I'm assuming (perhaps naively) that the default behaviour unless overridden by a system-wide policy would be to grant the container the privileges requested by the (trusted) publisher. Thus, a user would not ordinarily specify `--rights` via the CLI at all.

To enforce clarity between the two modes of operation we could disallow the use of `--rights` for "old" images and, conversely, the use of the existing options for "new" images.

WDYT?

**Comment by mato** [comment]
One more thing, orthogonal to my last comment, regarding naming, perhaps use `--rights=foo.bar,!bar.baz`. Or `--grant=foo.bar` and `--revoke=bar.baz`?

**Comment by thaJeztah** [comment]
related, older proposal https://github.com/moby/moby/pull/10712

**Comment by alexmavr** [comment]
@n4ss I don't see any mention of user namespaces for services in this model. Is this the right mechanism for supporting user namespaces for services, in equivalence to `docker create --userns=...`?

**Comment by n4ss** [comment]
POC and a more detailed and up-to-date version of the documentation. You can find it here:
Doc: https://docs.google.com/document/d/1j3BJUNBsgi-nxJHoIJHsXRRtVWT5lrwsI2EN9WMQaes/edit#heading=h.yhnr195944yh
POC: https://github.com/docker/libentitlement

user namespaces will be enabled through host.processes.none. The timeline for the entitlements project is not clear yet, we're currently proposing it as a common security effort for Moby and Kubernetes.

**Comment by marcellodesales** [comment]
@n4ss  Any update on this?

**Comment by n4ss** [comment]
@marcellodesales POC on `docker run` works, currently adding support in `docker service create` and api access control. Feel free to participate to and/or file issues (https://github.com/docker/libentitlement)

**Comment by uzzzerrrname** [comment]
Sorry, for dead-posting. Which is the target milestone for integration in Docker CE?
I assume it won't make it in 17.09, right?

**Comment by sudharkrish** [comment]
@n4ss Can you please let us know, if you have this support working for docker service create and api access control. Also, once docker service support is added, does the below command, allow access to all host devices(like /dev/device-name) on the host system in which the container is running.
docker service create --add-grant=host.devices.mount alpine

**Comment by n4ss** [comment]
@sudharkrish the POC works now with `docker service create`, more news very soon.

**Comment by ginjo** [comment]
@sudharkrish that's great news! Could you give us quick pionter on where to get the version with this new feature and the basics on how to use it with ```docker run```, ```docker service create```, and ```docker stack deploy```. Thanks.

**Comment by sudharkrish** [comment]
@n4ss Thanks for the info, please keep us posted.

**Comment by petrosagg** [comment]
The document mentions that during `docker build` a publisher can specify the entitlements required to run an image and that the system will automatically use those when running the image.

However, there is currently an assumption that if I download an image and run it nothing significantly bad will happen because it runs unprivileged by default.

Does this proposal address this by allowing the user that runs the image to provide a bounding set of entitlements under which the publisher can further constraint the execution? Otherwise it would be possible for a malicious publisher to publish an image that requires essentially `--privileged` level access and harm the system.

**Comment by n4ss** [comment]
@petrosagg i have to update the proposal but yes. The principle would be to specify how much you trust the baked-in profile. Providing one of of the following values: untrusted/trusted/unsure at start time would allow the user to be prompted to validate all/none/"above-average" permissions.

**Comment by westurner** [comment]
"ENH,SEC: Create additional sockets with limited permissions " https://github.com/moby/moby/issues/38879#issuecomment-472946314 lists two more granular potential sets of permissions for API access. Read only by default would be good.

> It's unfortunately common for projects to mount the docker socket into (privileged) containers. A configuration option for specifying additional sockets with limited privileges would be fairly simple and would make it easier to do it easily without doing it wrong.

**Comment by burnMyDread** [comment]
I have a question about the scope of these changes. When these changes are in place will I be able to set a default policy/set of entitlements? 

We would really like to have a way to set the default capabilities but if that is outside the scope of these changes I would be more than happy to open up a separate issue to cover that case.
